libraries/AP_Common/MultiHeap.cpp

```diff
@@ -0,0 +1,141 @@
+/*
+  multiple heap interface, allowing for an allocator that uses
+  multiple underlying heaps to cope with multiple memory regions on
+  STM32 boards
+ */
+
+#include <AP_HAL/AP_HAL.h>
+#include <AP_HAL/Util.h>
+#include <AP_Math/AP_Math.h>
+
+#include "MultiHeap.h"
+
+/*
+  on ChibiOS allow up to 4 heaps. On other HALs only allow a single
+  heap. This is needed as hal.util->heap_realloc() needs to have the
+  property that heap_realloc(heap, ptr, 0) must not care if ptr comes
+  from the given heap. This is true on ChibiOS, but is not true on
+  other HALs
+ */
+#if CONFIG_HAL_BOARD == HAL_BOARD_CHIBIOS
+#define MAX_HEAPS 4
```

Comment:
Can we get this wrapped with an `ifndef `so we can provide it via hwdef?

Reply:
done